File: easy/53_Maximum_Subarray.cpp
#include "../libstruct.h"

using namespace std;

/*
 * NO.53
 * Find the contiguous subarray within an array (containing at 
 * least one number) which has the largest sum.
 */

class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int ans = nums[0], sum = 0;
        for(int i = 0; i < nums.size(); i++){
            sum += nums[i];
            ans = max(sum, ans);
            sum = max(sum, 0);
        }
        return ans;
    }
};

//in: [-2,1,-3,4,-1,2,1,-5,4],
//out: sum = 6 ([4,-1,2,1])
int main(int argc, char const *argv[])
{
    Solution solution;
    vector<int> vec = {-2, 1, -3, 4, -1, 2, 1, -5, 4};
    cout << solution.maxSubArray(vec) << endl;
    return 0;
}

File: easy/53_Maximum_Subarray.swift
//
//  53_Maximum_Subarray.swift
//  LeetCode
//
//  Created by wangzhe on 2018/7/12.
//  Copyright © 2018年 wangzhe. All rights reserved.
//

import Foundation

/*
 * NO.53
 * Given an integer array nums, find the contiguous subarray (containing at least one number)
 * which has the largest sum and return its sum.
 */

class Solution {
    func maxSubArray(_ nums: [Int]) -> Int {
        guard nums.count > 0 else {
            return -1
        }
        var cur = 0, res = nums[0]
        for num in nums {
            if cur <= 0 {
                cur = num
            } else {
                cur += num
            }
            if cur > res {
                res = cur
            }
        }
        return res
    }
}
